### PR TITLE
OCPCLOUD-1250: add annotation to enable gpu autoscaling

### PR DESCRIPTION
**Pull request by SamuelStuchly**
Tested with `quay.io/elmiko/busybox` image with `limits: nvidia.com/gpu: 1` as workload for autoscaling tests.

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
just one question

File: pkg/cloud/gcp/actuators/machineset/controller.go [hunk 1]
```diff
@@ -156,6 +157,19 @@ func (r *Reconciler) reconcile(machineSet *machinev1.MachineSet) (ctrl.Result, e
 	machineSet.Annotations[cpuKey] = strconv.FormatInt(machineType.GuestCpus, 10)
 	machineSet.Annotations[memoryKey] = strconv.FormatInt(machineType.MemoryMb, 10)
 
+	if (len(providerConfig.GuestAccelerators)) == 0 {
+		if len(machineType.Accelerators) == 0 {
+			machineSet.Annotations[gpuKey] = strconv.FormatInt(0, 10)
+		} else {
+			// Accelerators will always be max size of 1
+			machineSet.Annotations[gpuKey] = strconv.FormatInt(machineType.Accelerators[0].GuestAcceleratorCount, 10)
```
[the remark this answers is not in the document]
Comment (SamuelStuchly) [reply pkg/cloud/gcp/actuators/machineset/controller.go:165]: Yes, the test case which was added is exactly for this code path.

[the remark this answers is not in the document]
Comment (elmiko) [reply pkg/cloud/gcp/actuators/machineset/controller.go:165]: ack, thanks. i was having difficulty picking it out of the tests.

File: pkg/cloud/gcp/actuators/machineset/controller.go [hunk 2]
```diff
@@ -156,6 +157,19 @@ func (r *Reconciler) reconcile(machineSet *machinev1.MachineSet) (ctrl.Result, e
 	machineSet.Annotations[cpuKey] = strconv.FormatInt(machineType.GuestCpus, 10)
 	machineSet.Annotations[memoryKey] = strconv.FormatInt(machineType.MemoryMb, 10)
 
+	if (len(providerConfig.GuestAccelerators)) == 0 {
```
Comment (JoelSpeed) [inline pkg/cloud/gcp/actuators/machineset/controller.go:160]: Nit, no need for extra parenthesis
```suggestion
	if len(providerConfig.GuestAccelerators) == 0 {
```

File: pkg/cloud/gcp/actuators/machineset/controller.go [hunk 3]
```diff
@@ -156,6 +157,19 @@ func (r *Reconciler) reconcile(machineSet *machinev1.MachineSet) (ctrl.Result, e
...
+			machineSet.Annotations[gpuKey] = strconv.FormatInt(0, 10)
+		} else {
+			// Accelerators will always be max size of 1
+			machineSet.Annotations[gpuKey] = strconv.FormatInt(machineType.Accelerators[0].GuestAcceleratorCount, 10)
+		}
+
+	} else {
+		// Guest accelerators will always be max size of 1
+		machineSet.Annotations[gpuKey] = strconv.FormatInt(providerConfig.GuestAccelerators[0].AcceleratorCount, 10)
+	}
```
Comment (JoelSpeed) [inline pkg/cloud/gcp/actuators/machineset/controller.go:171]: This seems quite hard to follow right now with these else statements, wonder if this might be easier to understand as a switch. Switches are ordered so the top case will take priority. WDYT?

```
switch {
case len(providerConfig.GuestAccelerators) > 0:
  machineSet.Annotations[gpuKey] = strconv.FormatInt(providerConfig.GuestAccelerators[0].AcceleratorCount, 10)
case len(machineType.Accelerators) > 0:
  machineSet.Annotations[gpuKey] = strconv.FormatInt(machineType.Accelerators[0].GuestAcceleratorCount, 10)
default:
  machineSet.Annotations[gpuKey] = strconv.FormatInt(0, 10)
}
```

File: pkg/cloud/gcp/actuators/machineset/controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/cloud/gcp/actuators/machineset/controller.go:165]: Do we have a test case for the other case?

[the remark this answers is not in the document]
Comment (SamuelStuchly) [reply pkg/cloud/gcp/actuators/machineset/controller.go]: We didnt, since it didnt seem necessary. It might be nice to have, so i added it.
